File: 프로그래머스/3/131703. 2차원 동전 뒤집기/2차원 동전 뒤집기.py
def solution(beginning, target):
    rlen = len(beginning)
    clen = len(beginning[0])
    answer = 10000000
    
    for mask in range(1 << rlen):
        flip_count = 0
        board = []
        
        for r in range(rlen):
            if mask & (1 << r):
                flip_count += 1
                board.append([(1 if x == 0 else 0) for x in beginning[r]])
            else:
                board.append(beginning[r][:])
        
        flag = True
        
        for j in range(clen):
            temp = board[0][j] == target[0][j]
            for i in range(rlen):
                if temp != (board[i][j] == target[i][j]):
                    flag = False
                    break
            
            if flag == False:
                break
            if temp == False:
                flip_count += 1
        
        if flag:
            answer = min(answer, flip_count)
            
    
    return -1 if answer == 10000000 else answer
                
        

File: 프로그래머스/3/131703. 2차원 동전 뒤집기/2차원 동전 뒤집기.js
function solution(beginning, target) {
    const diff = beginning.map((row,i) => row.map((val, j)=> val === target[i][j]));
    const rlen = beginning.length;
    const clen = beginning[0].length;
    let answer = Infinity;
    
    // 행의 경우의 수 (뒤집기 + 그대로 => 2 ** (행의 개수))
    for(let mask=0;mask<(1<<rlen);mask++){
        // 행의 뒤집힘에 따른 열의 뒤집힘 횟수 계산
        let flipCount = 0;
        const board = []
        
        for(let r=0;r<rlen;r++){
            if(mask & (1 << r)){ // r번째 행이 1이면 뒤집기
                flipCount++;
                board.push(beginning[r].map(val=> val === 0 ? 1 : 0))
            }else{
                board.push([...beginning[r]])
            }
        }
        
        let flag = true;
        
        for(let j=0;j<clen;j++){
            // 시작점이 같은지 다른지를 확인
            let temp = board[0][j] === target[0][j];
            for(let i=0;i<rlen;i++){
                if((board[i][j] === target[i][j]) !== temp){
                    flag = false;
                    break;
                }
            }
            
            if(!flag) break;
            if(!temp) flipCount++;
        }
        
        if(flag){
            answer = Math.min(answer, flipCount);
        }
    }

    
    return answer === Infinity ? -1 : answer
}